search: score non-main index entries lower

bradking opened the pull request:
Since #10819, index entries are returned as search results. When the query string exactly matches an indexed term, all index entries become search results with score 100.  This places them above most other search results.

Reporting "main" index entries early in the search results makes sense, but non-main index entries are often just arbitrary cross-references to the indexed term.  Give non-main entries lower scores so they are not ordered early in search results.  This avoids obscuring the main entries and other more-important matches such as document titles.

Fixes: #11578

bradking committed:
search: score non-main index entries lower

Since commit 8ae81839b (Support searching for index entries (#10819),
2022-09-20, v5.2.0~16), index entries are returned as search results.
When the query string exactly matches an indexed term, all index
entries become search results with score 100.  This places them above
most other search results.

Reporting "main" index entries early in the search results makes sense,
but non-main index entries are often just arbitrary cross-references to
the indexed term.  Give non-main entries lower scores so they are not
ordered early in search results.  This avoids obscuring the main entries
and other more-important matches such as document titles.

Fixes: #11578

picnixz commented on sphinx/search/__init__.py:
```diff
@@ -396,10 +396,10 @@ def freeze(self) -> dict[str, Any]:
                 alltitles.setdefault(title, []).append((fn2index[docname], titleid))
 
-        index_entries: dict[str, list[tuple[int, str]]] = {}
+        index_entries: dict[str, list[tuple[int, str, str]]] = {}
```
```suggestion
        index_entries: dict[str, list[tuple[int, str, bool]]] = {}
```

picnixz commented on sphinx/search/__init__.py:
```diff
@@ -396,10 +396,10 @@ def freeze(self) -> dict[str, Any]:
         for docname, entries in self._index_entries.items():
             for entry, entry_id, main_entry in entries:
-                index_entries.setdefault(entry.lower(), []).append((fn2index[docname], entry_id))
+                index_entries.setdefault(entry.lower(), []).append((fn2index[docname], entry_id, main_entry))
```
```suggestion
                index_entries.setdefault(entry.lower(), []).append((fn2index[docname], entry_id, main_entry == 'main'))
```

picnixz commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,11 @@ const Search = {
       if (entry.includes(queryLower) && (queryLower.length >= entry.length/2)) {
-        for (const [file, id] of foundEntries) {
-          let score = Math.round(100 * queryLower.length / entry.length)
+        for (const [file, id, main] of foundEntries) {
```
```suggestion
        for (const [file, id, isMain] of foundEntries) {
```

picnixz commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,11 @@ const Search = {
+        for (const [file, id, main] of foundEntries) {
+          // Give main entries a high score so they are ordered early.
+          // Give non-main entries a low score so they are ordered later.
+          let maxScore = (main === "main" ? 100 : 1);
```
```suggestion
          const maxScore = isMain ? 100 : 1;
```

picnixz commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,11 @@ const Search = {
+          // Give main entries a high score so they are ordered early.
+          // Give non-main entries a low score so they are ordered later.
+          let maxScore = (main === "main" ? 100 : 1);
+          let score = Math.round(maxScore * queryLower.length / entry.length);
```
```suggestion
          const score = Math.round(maxScore * queryLower.length / entry.length);
```

picnixz commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,11 @@ const Search = {
-          let score = Math.round(100 * queryLower.length / entry.length)
+        for (const [file, id, main] of foundEntries) {
+          // Give main entries a high score so they are ordered early.
+          // Give non-main entries a low score so they are ordered later.
```
Maybe add a reminder that since `queryLower.length >= entry.length/2`, having a maxScore of `1` (any choice of `< 50` would actually work) for non-main entries still guarantees that we don't have some weird cases where a non-main entry has a score higher than a main entry. 

Indeed, the *min* score of a main-entry is 50 and thus, it suffices for a non-main entry to have a *max* score of 50 and we're done.

bradking replied:
I extended the comment for this, thanks.  I also applied the other suggestions.

picnixz commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,17 @@ const Search = {
+          // score of 50.  We can choose any weight lower than 50 for non-main
+          // entries to ensure they can never be ordered before main entries.
+          // Since non-main entries are typically just arbitrary cross-references
+          // to the indexed entity, give them a very low score.
```
I'm a bit nitpicky here may I suggest the following?

```suggestion
          // Give main entries a high score so they are ordered early
          // and non-main entries a low score so they are ordered later.
          //
          // We only get here if the query length is at least half the entry
          // length, hence, with a weight of 100, main entries have a minimum
          // score of 50.  We can choose any weight lower than 50 for non-main
          // entries to ensure they are always ordered after main entries, but
          // since they are typically just arbitrary cross-references to the 
          // indexed entity, we choose a very low score.
```

bradking replied:
Applied, thanks.

bradking committed:
search: polish comment explaining non-main index entry scoring

jayaddison commented on sphinx/themes/basic/static/searchtools.js:
```diff
@@ -308,8 +308,18 @@ const Search = {
+          // since they are typically just arbitrary cross-references to the
+          // indexed entity, we choose a very low score.
+          const maxScore = isMain ? 100 : 1;
+          const score = Math.round(maxScore * queryLower.length / entry.length);
```
Given the rounding involved, how different is this to setting the score to a fixed value (`0` or `null` for example) for all of the non-index entries?

Maths tricks in scoring can be a useful way to achieve re-ordering, but they don't compose well - so once we begin using one or two of them, it'll become much more difficult to use scoring for any other adjustments.

(my preference would be to avoid the use of mathematical separation of the non-index entries, but to find some other technique, such as appending them separately after the rest of the documents are scored.  it seems debatable though)

picnixz replied:
Oh good catch. I totally missed that point. Indeed, we should perhaps change the maxScore for non-main entries to 49, so that we have a better approximation.

bradking replied:
I tried 49 locally, but in my use case it still ordered non-main entries before more relevant search results.

I updated this PR to use 10 instead.

> avoid the use of mathematical separation

I've opened #11696 to do that instead.

bradking committed:
search: score non-main index entries high enough for internal ordering

Previously we gave non-main index entries a maximum score of 1.
After rounding, that becomes either 0 or 1.  Raise their score
to 10 so there is some room for ordering among them.
